src/platform/backends/hyperv_api/hcs/hyperv_hcs_schema_version.cpp

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright (C) Canonical, Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+#include <hyperv_api/hcs/hyperv_hcs_schema_version.h>
+
+#include <hyperv_api/hyperv_api_string_conversion.h>
+
+#include <platform/platform_win.h>
+
+#include <optional>
+
+namespace
+{
+
+// https://www.wikiwand.com/en/articles/Windows_10_version_history
+// https://www.wikiwand.com/en/articles/Windows_11_version_history
+// https://www.wikiwand.com/en/articles/List_of_Microsoft_Windows_versions
+enum class WindowsBuildNumbers : std::uint32_t
+{
+    // April 2018 Update, April 30, 2018
+    win10_1809 = 17763,
+    // May 2019 Update, May 21, 2019
+    win10_19H1 = 18362,
+    // May 2020 Update, May 27, 2020
+    win10_20H1 = 19041,
+    // Codename "Vibranium", August 18, 2021
+    srv22_21H2 = 20348,
+    // Codename "Sun Valley", October 5, 2021
+    win11_21H2 = 22000
+};
+} // namespace
+
+namespace multipass::hyperv::hcs
+{
+
+SchemaUtils::SchemaUtils(const Singleton<SchemaUtils>::PrivatePass& pass) noexcept
+    : Singleton<SchemaUtils>::Singleton{pass}
+{
+}
+
+// ---------------------------------------------------------
+
+HcsSchemaVersion SchemaUtils::get_os_supported_schema_version() const
+{
+    const static auto cached_schema_version = []() -> std::optional<HcsSchemaVersion> {
+        if (const auto winver = platform::get_windows_version())
+        {
+            struct SchemaVersionBuildNumberMapping
+            {
+                HcsSchemaVersion version;
+                WindowsBuildNumbers required_build_number;
+            };
+
```

Comment:
Could this go into the anonymous namespace, preferably along with an `operator>` that you could pass to sort? That would lighten this function a little.